File: leetcode/boats-to-save-people.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        int boat_count = 0;
        
        sort(people.begin(), people.end());
        
        auto iterator = people.begin();
        
        while(iterator != people.end()) {
            auto scan_iterator = --people.end();
            while (scan_iterator != iterator) {
                if (*iterator + *scan_iterator <= limit) {
                    people.erase(scan_iterator);
                    break;
                }
                --scan_iterator;
            }
            
            ++boat_count;
            ++iterator;
        }
        
        return boat_count;
    }
};

File: leetcode/boats-to-save-people.py
from typing import List


class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()

        l = 0
        r = len(people) - 1

        boats = 0

        while l <= r:
            if people[r] == limit or people[r] + people[l] > limit:
                r -= 1
            else:
                l += 1
                r -= 1

            boats += 1

        return boats
